modules/dcache/src/main/java/org/dcache/services/ssh2/Ssh2Admin.java

```diff
@@ -318,6 +331,71 @@ public boolean authenticate(String userName, PublicKey key,
         }
     }
 
+    private static class RemoteHostValidator {
+
+        private enum Outcome {
+            ALLOW, DENY, DEFER
+        }
+
+        private boolean isValidHost(String line, ServerSession session) {
+            for (String linePart : Splitter.on(" ").trimResults().omitEmptyStrings().split(line)) {
+                if (linePart.startsWith("from=") && linePart.endsWith("\"")) {
+                    Set<Outcome> outcomes = EnumSet.noneOf(Outcome.class);
+                    for (String pattern : Splitter.on(",").trimResults().omitEmptyStrings().split(strip(linePart))) {
+                        outcomes.add(patternMatchesHost(pattern, session));
+                    }
+                    return outcomes.contains(Outcome.ALLOW) && !outcomes.contains(Outcome.DENY);
+                }
+            }
+            return true;
+        }
+
+        private Outcome patternMatchesHost(String pattern, ServerSession session) {
+            boolean patternMatches = false;
+            boolean patternIsNegated = pattern.startsWith("!");
+            if (patternIsNegated) {
+                pattern = pattern.substring(1);
+            }
+            SocketAddress remote = session.getClientAddress();
+            if (remote instanceof InetSocketAddress) {
+                String remoteAddress = ((InetSocketAddress) remote).getAddress().getHostAddress();
+                try {
+                    patternMatches = addressInCidrRange(pattern, remoteAddress);
+                } catch (IllegalArgumentException e) {
+                    String remoteName = ((InetSocketAddress) remote).getHostName();
+                    if (Glob.isGlob(pattern)) {
+                        Pattern p = Glob.parseGlobToPattern(pattern);
+                        Matcher addressMatcher = p.matcher(remoteAddress);
+                        Matcher nameMatcher = p.matcher(remoteName);
+                        patternMatches = (addressMatcher.matches() || nameMatcher.matches());
+                    } else {
+                        patternMatches = pattern.equals(remoteName);
+                    }
+                }
+            }
+            if (patternIsNegated && patternMatches) {
```

Comment:
Although the code is OK, as-is, I think it could be simplified by removing the `patternMatches` variable and adding a variable that holds the response to match; something like:

    Outcome matchOutcome = patternisNegated ? Outcome.DENY : Outcome.ALLOW;

Then, the code can simply `return matchOutcome;` when there's a match.

Technically, it would also be possible to update the patch to eliminate the patternIsNegated variable, too.  I'm not sure if that improves or detracts from the code's readability.

In any case, have a play with these ideas and see if you like them.